Reworking Range Widget

SaumiaSinghal opened the pull request:
This PR includes rewriting the test coverage and class of Range Picker Widget 

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
I verified the behavior on the `master` branch and on this branch for the attached form. I didn't find any behavioral changes in the widget. I updated to the latest google material library version 1.2.1 to display Range Widgets using Material Slider.

There is a slight clipping of the ripple effect for the vertical slider, but I found the same on the master branch as well. @seadowg and I decided to leave that for now. I will keep looking for solutions.

#### Do we need any specific form for testing your changes? If so, please attach one.
Form having both Range Integer and Decimal widgets like 
[Test Range Picker Widgets.xml.txt](https://github.com/getodk/collect/files/4917565/Test.Range.Picker.Widgets.xml.txt)
[Test Range Widget.xml.txt](https://github.com/getodk/collect/files/4931460/Test.Range.Widget.xml.txt)

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

SaumiaSinghal committed:
update range widget xml file

SaumiaSinghal committed:
fix slider.setValue() behaviour

SaumiaSinghal committed:
fix slider colors

SaumiaSinghal committed:
update range integer java class

SaumiaSinghal committed:
update range integer java class

SaumiaSinghal committed:
add check for vertical appearnce

SaumiaSinghal committed:
add unit tests

SaumiaSinghal committed:
make getAnswerShouldReflectActualValueSetViaSeekBar() test greeen

SaumiaSinghal committed:
update slider color to theme colors

SaumiaSinghal committed:
add unit test for range decimal widget

SaumiaSinghal committed:
update range decimal widgte java class

SaumiaSinghal committed:
create RangeWidgetUtils class

SaumiaSinghal committed:
add unit tests for RangeWidgetUtils

SaumiaSinghal committed:
create Range picker classes

SaumiaSinghal committed:
update RangePickerIntegerWidget

SaumiaSinghal committed:
create CustomRangeSlider class and code refactor

SaumiaSinghal committed:
add unit tests for CustomRangeSlider

SaumiaSinghal committed:
code refactor

SaumiaSinghal committed:
create custom nested class for returning range widget layout elements in RangeWidgetUtils

SaumiaSinghal committed:
fix usingReadOnlyOption_disablesTheSlider unit test

SaumiaSinghal committed:
fix check_quality error

SaumiaSinghal committed:
code refactor

SaumiaSinghal committed:
code refactor

SaumiaSinghal committed:
unit tests refactor

SaumiaSinghal committed:
add unit tests for RangePickerIntegerWidget

SaumiaSinghal committed:
add unit tests

SaumiaSinghal committed:
add unit tests for RangePickerDecimalWidget

SaumiaSinghal committed:
update RangePickerDecimalWidget class

SaumiaSinghal committed:
code refactor to make unit tests pass

SaumiaSinghal committed:
refactor code

SaumiaSinghal committed:
add unit tests for RangeWidgetUtils

SaumiaSinghal commented on collect_app/src/test/java/org/odk/collect/android/widgets/RangePickerDecimalWidgetTest.java:
```diff
@@ -0,0 +1,147 @@
+
+        assertNotNull(numberPickerDialog);
+    }
+
```
@seadowg `clickingPickerButton_showsNumberPickerDialog()` unit test is failing. I'm looking onto it. If you have a hint, please let me know.

seadowg replied:
Yeah if you debug you'll see that in `RangeWidgetUtils.showNumberPickerDialog` the `catch` branch will run as there is a `ClassCastException`. The code is trying to cast the `context` to a `FormEntryActivity` but it doesn't need to be one of those to add a `Fragment` right? I think it just needs to cast it to a `FragmentActivity` (`Activity` doesn't have `getSupportFragmentManager` on it).

I think you'll also need to have the activity in fully created state to add a fragment (rather than one that's just been "built").  Maybe just for those tests you can create a different test activity using `createThemedActivity`. The extra work is pretty slow so good to only use it in tests that need it!

SaumiaSinghal replied:
Right I think that's why the tests simply failed instead of giving an error

SaumiaSinghal committed:
shift checkNumberPickerDialogShown test to RangeWidgetUtilsTest

SaumiaSinghal committed:
implement conitnuos slider based on 'no ticks' appearance

SaumiaSinghal committed:
fix failing unit tests

SaumiaSinghal committed:
fix crash when start value is greater tahn end value

SaumiaSinghal committed:
add touch listener on the slider

SaumiaSinghal committed:
add unit tests for TrackingTouchSlider

SaumiaSinghal committed:
fix failing unit tests

seadowg reviewed (changes requested):
The Espresso tests are not compiling. I know the tests are slow and hard to run for every change but it's a good idea to at least be running `AllWidgetsFormTest` (as it's pretty relevant to the widgets) before marking these PRs as ready for review.

SaumiaSinghal commented:
@seadowg! Everything seems fine now. I'll check the espresso tests as well from now on.

SaumiaSinghal committed:
add unit tests

SaumiaSinghal committed:
use material slider instead of seekbar

SaumiaSinghal committed:
implement Slider.onTouchListener

SaumiaSinghal committed:
remove RangeWidget and RangeWidgetTest class

SaumiaSinghal committed:
add unit tests for vertical slider

SaumiaSinghal committed:
update RangePickerIntegerWidget class

seadowg commented on collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java:
```diff
@@ -2633,8 +2634,12 @@ public void onSaveFormIndexError(String errorMessage) {
             for (QuestionWidget qw : ((ODKView) currentView).getWidgets()) {
-                if (qw instanceof RangeWidget && widgetId == qw.getId()) {
-                    ((RangeWidget) qw).setNumberPickerValue(value);
+                if (qw instanceof RangePickerIntegerWidget && widgetId == qw.getId()) {
```
I'm not a big fan of the `FormEntryActivity` having to know about widget implementations here but this structure existed a long time before this PR and I think it's be awkward to unravel.

SaumiaSinghal replied:
@seadowg I think the other option that we can use is having a separate `viewModel` for `RangeWidget`. I also don't like having a callback implemented in `FormEntryActivity`.

seadowg commented on collect_app/src/test/java/org/odk/collect/android/views/TrackingTouchSliderTest.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    @Test
+    public void onStopTrackingTouch_doesNotSuppressFlingGesture() {
```
Probably worth combining this with the test above so it tests that the `ACTION_UP` **after** the `ACTION_DOWN` works. With these tests the implementation could just set `isTrackingTouch` to `true` on an `ACTION_DOWN` and then never change it back 😉

seadowg commented on collect_app/src/test/java/org/odk/collect/android/views/TrackingTouchSliderTest.java:
```diff
@@ -0,0 +1,45 @@
+    @Before
+    public void setUp() {
+        ApplicationProvider.getApplicationContext().setTheme(R.style.Theme_Collect_Light);
+
```
Oh this is very cool! It never occurred to me to just set the application context theme. This could probably help us speed up tests where we're building an `Activity` just to set the context theme. Maybe you could add this to `RobolectricHelpers` as `createThemedContext(int theme)` so we  can use it other places?

SaumiaSinghal replied:
Yes sure @seadowg :)

seadowg commented on collect_app/src/main/java/org/odk/collect/android/widgets/RangeDecimalWidget.java:
```diff
@@ -18,20 +18,56 @@
+    @Override
+    protected View onCreateAnswerView(Context context, FormEntryPrompt prompt, int answerFontSize) {
+        rangeQuestion = (RangeQuestion) getFormEntryPrompt().getQuestion();
+        RangeWidgetUtils.RangeWidgetLayoutElements layoutElements = RangeWidgetUtils.setUpLayoutElements(context, prompt);
```
I think in future this could probably be extracted to its own custom view but for the moment it's great progress!

SaumiaSinghal replied:
Yes correct @seadowg Thank you!

seadowg commented on collect_app/src/main/java/org/odk/collect/android/widgets/UrlWidget.java:
```diff
@@ -43,18 +44,20 @@ public UrlWidget(Context context, QuestionDetails questionDetails, CustomTabHelp
     @Override
     protected View onCreateAnswerView(Context context, FormEntryPrompt prompt, int answerFontSize) {
-        binding = UrlWidgetAnswerBinding.inflate(((Activity) context).getLayoutInflater());
+        binding = WidgetAnswerBinding.inflate(((Activity) context).getLayoutInflater());
```
I think this layout should be called `button_and_answer` or something. I don't like suggesting that it's the "default" answer view.

SaumiaSinghal replied:
I renamed it to `widget_button_and_answer_layout`

seadowg commented on collect_app/src/main/res/layout/range_widget_horizontal.xml:
```diff
@@ -31,16 +34,15 @@ limitations under the License.
-            android:theme="@style/Base.Widget.AppCompat.SeekBar.Discrete" />
+            app:thumbColor="?colorSecondary"
+            app:tickColor="?colorOnBackground"
+            app:haloColor="?dividerColor"
```
I don't think we should use `dividerColor` here. Could you find a color to use in the standard [material color set](https://material.io/develop/android/theming/color) and use that instead? If it isn't in our themes already you might have to define it. 

If it seems doable you could then replace `dividerColor` everywhere and delete it? No worries if that seems like a distraction though.

SaumiaSinghal replied:
Yes sure

seadowg commented on collect_app/src/main/res/layout/range_widget_vertical.xml:
```diff
@@ -33,22 +34,18 @@ limitations under the License.
             android:layout_width="330dp"
             android:layout_height="wrap_content"
+            app:thumbColor="?colorSecondary"
+            app:tickColor="#696969"
```
This should use theme colors rather than hardcoded colors.

seadowg reviewed (changes requested):
Just a few changes but looking great!

SaumiaSinghal committed:
code refactor

SaumiaSinghal committed:
code refactor

SaumiaSinghal committed:
used material design color

SaumiaSinghal committed:
create method createThemedContext() in RobolectricHelpers

SaumiaSinghal committed:
code refactor

seadowg commented on collect_app/src/main/res/values/attrs.xml:
```diff
@@ -2,6 +2,7 @@
 <resources>
     <declare-styleable name="Collect">
         <attr name="iconColor" format="color" />
+        <attr name="sliderTrackColor" format="color" />
```
I think you misunderstood what I was trying to communicate before. I just meant if we were missing one of the standard material theme colors we might have to add it to our themes. Sorry if that was confusing! 

The theme shouldn't have to know that there is a slider that needs a color right? The slider should use a material theme color we have a value for not a new custom one.

SaumiaSinghal replied:
Oh now I get it. We have values for almost all the theme colors except for `colorBackground`. However, I feel that I shouldn't be using it as `sliderTrackColor`, as it doesn't sound correct. I feel like I should use `colorOnBackground` as slider track color. Am I correct? I found that we are not using the value of  `colorOnBackground` anywhere in the app. Should I update it?

seadowg replied:
I'd say `colorOnSurface` probably makes sense?

seadowg commented on collect_app/src/main/java/org/odk/collect/android/fragments/dialogs/NumberPickerDialog.java:
```diff
@@ -55,12 +53,7 @@ public static NumberPickerDialog newInstance(int widgetId, String[] displayedVal
-        } catch (ClassCastException e) {
-            throw new ClassCastException(context.toString() + " must implement OnHeadlineSelectedListener");
-        }
+        rangePickerViewModel = new ViewModelProvider(requireActivity(), new ViewModelProvider.NewInstanceFactory()).get(RangePickerViewModel.class);
```
I don't think you need to pass a `NewInstanceFactory` here. Could you not just use the 1 arg constructor for `ViewModelProvider`?

seadowg commented on collect_app/src/main/java/org/odk/collect/android/widgets/RangePickerDecimalWidget.java:
```diff
@@ -0,0 +1,100 @@
+
+        rangePickerViewModel.getNumberPickerValue().observe(((ScreenContext) getContext()).getViewLifecycle(), answer -> {
+            if (answer != null) {
+                progress = rangePickerViewModel.getNumberPickerProgress(rangeStart, rangeStep, rangeEnd, answer);
```
I feel like `getNumberPickerProgress` should probably remain a util method as it doesn't rely on any state in the view model.

SaumiaSinghal committed:
remove slider track color and use color_on_surface_low_emphasis

seadowg commented on collect_app/src/test/java/org/odk/collect/android/widgets/RangePickerDecimalWidgetTest.java:
```diff
@@ -0,0 +1,128 @@
+        int index = 0;
+
+        for (double i = 5.5F; i >= 1.5F; i -= 0.5F, index++) {
+            assertThat(widget.displayedValuesForNumberPicker[index], equalTo(String.valueOf(i)));
```
I'm not a fan of testing this field here as there is no reason for it to be public. Really the test should click on the button and test that the dialog receives the right values. I know you're reworking how all that functions though so I think we go with this for the moment!

As a general tip for tests like this I would avoid the `for` loop and just write out what I expect the values to be. That way it's clearer to the reader of the test. Tests serve as documentation as much as they serve as checks so it's good for them to be very clear what the behaviour they expect is.

seadowg reviewed (changes requested):
Swiping the slider all the way to left or right cuts of the grey highlight circle. I think we should work out if we can fix that before movign this to QA.

SaumiaSinghal committed:
fix slider ripple

mmarciniak90 commented:
@SaumiaSinghal take a look at mentioned issues

|Description |released v1.27.3|PR|
|---|---|---|
|When you open Range widgets it looks like an answer is added because a blue dot is visible|![Screenshot_20200807-164529](https://user-images.githubusercontent.com/3907554/89657918-242ccd80-d8ce-11ea-963a-61f042c410e7.png)|![Screenshot_2020-08-07-16-48-34-217_org odk collect android](https://user-images.githubusercontent.com/3907554/89657921-24c56400-d8ce-11ea-9000-022c779a0760.jpg)|
|Read Only question is not grey out. There is no difference between editable and not editable question. |![Screenshot_20200807-164753](https://user-images.githubusercontent.com/3907554/89658146-740b9480-d8ce-11ea-9323-646a34872a16.png)|![Screenshot_2020-08-07-16-51-10-164_org odk collect android](https://user-images.githubusercontent.com/3907554/89658128-6eae4a00-d8ce-11ea-8e66-449b37e0100e.jpg)|
|Read only question with filled answer is not grey out |![Screenshot_20200807-170356](https://user-images.githubusercontent.com/3907554/89660003-001ebb80-d8d1-11ea-8bb7-1dc3e4d553b4.png)|![Screenshot_2020-08-07-17-02-48-815_org odk collect android](https://user-images.githubusercontent.com/3907554/89660015-04e36f80-d8d1-11ea-82b5-657beee5d3f6.jpg)|
|I'm not sure about showing the inactive button for the read-only picker. Sorry If I missed this requirement|![Screenshot_20200807-164921](https://user-images.githubusercontent.com/3907554/89658278-a4533300-d8ce-11ea-9958-990c843e8407.png)|![Screenshot_2020-08-07-16-52-26-927_org odk collect android](https://user-images.githubusercontent.com/3907554/89658287-a7e6ba00-d8ce-11ea-8f17-eb3866fec7be.jpg)|

mmarciniak90 commented:
@getodk-bot unlabel "needs testing"

SaumiaSinghal committed:
hide thumb when widget has no answer

SaumiaSinghal committed:
hide picker button when readonly and code refactor

SaumiaSinghal committed:
create slider tint list

SaumiaSinghal committed:
fix slider thumb and code refactor

SaumiaSinghal commented:
Hi @mmarciniak90 I fixed the mentioned issues with the Ui of the disabled slider and the picker button. Can you please test again?

kkrawczyk123 commented:
@SaumiaSinghal I can see another issue connected to dot displayed on the axis without choosing any answer when I follow scenario:
1. Open any range widget with an axis
2. Do not choose any number
3. Rotate screen - unexpected dot appears on the axis and the lowest value is selected automatically
Dot is still visible after rotating back to portrait.
![Screenshot (Aug 18, 2020 12_02_11 PM)](https://user-images.githubusercontent.com/33342488/90500421-23f0c580-e14b-11ea-8305-3fcb04dfdcdf.png)

@getodk-bot unlabel "needs testing"

SaumiaSinghal committed:
retain null answer on screen rotation

SaumiaSinghal commented:
Hi @kkrawczyk123! Thanks for reporting the issue. I fixed it. Can you test again please?

SaumiaSinghal committed:
create slider theme

SaumiaSinghal committed:
add unit tests

grzesiek2010 commented:
@SaumiaSinghal does this pr fix https://github.com/getodk/collect/issues/4030 too?

SaumiaSinghal commented:
Hi @grzesiek2010! This PR https://github.com/getodk/collect/pull/3995 fix #4030. It involves work on top of this PR.

kkrawczyk123 commented:
@SaumiaSinghal I have noticed an issue when select RTL language and then rotate the screen, axis in LTR is visible after screen rotation, see the gif: 
![ezgif com-video-to-gif (35)](https://user-images.githubusercontent.com/33342488/94000130-02f94100-fd97-11ea-8332-1c4fa5e50672.gif)

@getodk-bot unlabel "needs testing"

SaumiaSinghal commented:
Thanks @kkrawczyk123 for reporting the issue. Sorry for the late reply, I somehow missed the message. It seems there was an issue that the RTL language was getting reset on the branch on screen orientation. I rebased with the master branch and everything works fine now. Also, I updated google material library to a more stable and latest version 1.2.1.

kkrawczyk123 commented:
@getodk-bot label "needs testing"

kkrawczyk123 commented:
**Tested with success!**
Verified on Androis: 5.1, 7.1 and 9.0.

Tested scenarios:
- light/dark mode
- RTL lamguage
- rotate
- minimize 
- going backward/forward
- readonly
- jump from hierarchy 
- remove response 
- dont keep activities turned on/off

mmarciniak90 commented:
Verified also on Android versions: 7.0, 8.1, 10.0

@getodk-bot unlabel "needs testing"
@getodk-bot label "behavior verified"

grzesiek2010 commented:
@SaumiaSinghal please fix the conflicts

SaumiaSinghal committed:
resolve conflict in build.gradle

SaumiaSinghal committed:
refactor SuppressFlingGestureSlider

SaumiaSinghal committed:
fix slider behaviour when range start is greater than range end value

SaumiaSinghal committed:
refactor code

SaumiaSinghal committed:
code refactor

SaumiaSinghal committed:
code refactor

SaumiaSinghal committed:
refactor WidgetButtonAndAnswerBinding

SaumiaSinghal committed:
add styles for widget button and answer text view

lognaturel commented:
Thanks so much for getting this over the finish line, @SaumiaSinghal! 🎉 

There's a little bit of risk here since we can't easily track the conflict resolution through the rebase but any issues would be caught on usage or at least on regression testing.

I went back and forth on whether to squash. I think there are several commits that do provide hints as to decisions made so I decided to keep them all.
